### ISSUE TITLE
Allow reading tables in an untyped way

### ISSUE THREAD

**Issue by holly-hacker** [cberner/redb#741]
I'm using redb in a hobby project and while I appreciate its typed tables, I also wish to be able to list/dump the contents of every table without knowing its type at compile time. Ideally, I'd be able to iterate over every table as if it's a `BTreeMap<&[u8], &[u8]>`, similar to sled's [`sled::Db::export(&self)`](https://docs.rs/sled/latest/sled/struct.Db.html#method.export).

I noticed that the current `master` branch contains a `redb::ReadOnlyUntypedTable` type which seems to be what I want, but it only allows for getting stats and not the contents of the table.

**Comment by cberner** [comment]
Can you tell me more about your use case? I considered adding that, but
decided not to since the bytes are meaningless without the type

**Comment by holly-hacker** [comment]
Some usecases I can think off:
- more easily migrating to another kv store that doesn't use typed collections/namespaces
- creating backups that don't rely on redb to be read correctly, and would thus not be influenced by possible bugs
- creating a tool to inspect the database without having to hardcode support for each table individually (this is what I was trying to do)

**Comment by cberner** [comment]
I see. For the first two, I think that would be best built into the
application and so the type information would be available.

Ya, the third makes sense, but I don't want to support it in the public API
because it would make the binary representation of types part of the public
API.

It's probably not too hard to add that, if you fork it and remove the type
checks

**Comment by cberner** [comment]
I don't think that would work because there are user defined types
(RedbValue is public)

**Comment by holly-hacker** [comment]
Ah, I figured it'd work since TableDefinition seems to work in roughly the same way.

Feel free to close this issue if you think this isn't going to happen. I can currently continue with my workaround so this isn't blocking me.

**Comment by cberner** [comment]
Nope, even `TableDefinition` needs to know the type at compile time. You just don't need to specify them in many cases because of type inference

**Comment by holly-hacker** [comment]
fwiw, the type in my example *is* known at compile time: `key.get::<u32>()`

**Comment by cberner** [comment]
Oh, if you have the types at compile time then why doesn't the normal
open_table() and get() API work for you?

**Comment by holly-hacker** [comment]
I wouldn't have the types themselves, but you can write code that checks what the type is at runtime to implement specialized read or write logic:
```rs
fn display(entry: &UntypedValue) -> String {
    if let Some(entry) = entry.get::<u32>() {
        return format!("{entry}");
    }
    if let Some(()) = entry.get::<()>() {
        return "unit".to_string();
    }
    if let Some(entry) = entry.get::<&'static str>() {
        return entry.into();
    }
    // ... more
    return "<unsupported type>".into();
}
```
This example shows displaying a value, but you could implement editing this way too. It could also be useful to expose a TypeId or type name or something to make this matching easier.

The current `TypeDefinition`/`open_table()` api wouldn't work here because it requires specifying the correct type for both the key and value, so if you want to implement support for `n` types you would need `n^2` cases to check since you need to guess both types correctly.

**Comment by cberner** [comment]
You don't need to guess at all n^2 types. You'll receive a
TableTypeMismatch error which contains the type information

**Comment by holly-hacker** [comment]
That could work, though probably not as conveniently as an untyped value API. I'll have to play around with it some time, but the workaround I have works for now.
